src/coreclr/jit/importer.cpp

```diff
@@ -2483,6 +2483,32 @@ GenTree* Compiler::impTypeIsAssignable(GenTree* typeTo, GenTree* typeFrom)
     return nullptr;
 }
 
+GenTree* Compiler::impGetGenericTypeDefinition(GenTree* type)
+{
+    // This intrinsic requires the first arg to be some `typeof()` expression,
+    // ie. it applies to cases such as `typeof(...).GetGenericTypeDefinition()`.
+    CORINFO_CLASS_HANDLE hClassType = NO_CLASS_HANDLE;
+    if (gtIsTypeof(type, &hClassType))
+    {
+        // Check that the 'typeof()' expression is being used on a type that is in fact generic.
+        // If that is not the case, we don't expand the intrinsic. This will end up using
+        // the usual Type.GetGenericTypeDefinition() at runtime, which will throw in this case.
+        if (info.compCompHnd->getTypeInstantiationArgument(hClassType, 0) != NO_CLASS_HANDLE)
+        {
+            CORINFO_CLASS_HANDLE hClassResult = info.compCompHnd->getTypeDefinition(hClassType);
+
+            GenTree* retNode = gtNewIconEmbClsHndNode(hClassResult);
+
+            // Drop the typeof(T) node
+            impPopStack();
```

Comment:
Hitting this assert:

> Assert failure(PID 31068 [0x0000795c], Thread: 24116 [0x5e34]): Assertion failed '(genActualType(op2->TypeGet()) == genActualType(info.compRetType)) || ((op2->TypeGet() == TYP_I_IMPL) && (info.compRetType == TYP_BYREF)) || ((op2->TypeGet() == TYP_BYREF) && (info.compRetType == TYP_I_IMPL)) || (varTypeIsFloating(op2->gtType) && varTypeIsFloating(info.compRetType)) || (varTypeIsStruct(op2) && varTypeIsStruct(info.compRetType)) : Possibly bad IL with CEE_ret at offset 0022h (op1=NULL op2=long stkDepth=0)' in 'Foo:Test1[System.Collections.Generic.KeyValuePair`2\[System.__Canon,int]]():System.Type' during 'Importation' (IL size 35; hash 0xec731554; FullOpts)

This `impPopStack` seems a bit sus, but it also does look correct to me. Looking into it...

Reply:
Reproes with:

```csharp
Test1<KeyValuePair<string, int>>();

[MethodImpl(MethodImplOptions.NoInlining)]
public static Type? Test1<T>() => typeof(T).GetGenericTypeDefinition();
```

Reply:
Was missing a handle -> `Type` conversion, fixed in 2a88644bdf18abc94a49262962b858eccb6cb972.